To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

Unsloth now supports Text-to-Speech (TTS) models. Read our [guide here](https://docs.unsloth.ai/basics/text-to-speech-tts-fine-tuning).

Read our **[Qwen3 Guide](https://docs.unsloth.ai/basics/qwen3-how-to-run-and-fine-tune)** and check out our new **[Dynamic 2.0](https://docs.unsloth.ai/basics/unsloth-dynamic-2.0-ggufs)** quants which outperforms other quantization methods!

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth vllm
else:
    # [NOTE] Do the below ONLY in Colab! Use [[pip install unsloth vllm]]
    !pip install --no-deps unsloth vllm==0.8.5.post1

In [2]:
!pip install wandb --quiet

In [ ]:
import wandb
wandb.login()

In [4]:
#@title Colab Extra Install { display-mode: "form" }
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth vllm
else:
    !pip install --no-deps unsloth vllm==0.8.5.post1
    # [NOTE] Do the below ONLY in Colab! Use [[pip install unsloth vllm]]
    # Skip restarting message in Colab
    import sys, re, requests; modules = list(sys.modules.keys())
    for x in modules: sys.modules.pop(x) if "PIL" in x or "google" in x else None
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft "trl==0.15.2" triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf "datasets>=3.4.1" huggingface_hub hf_transfer
    !pip install transformers==4.51.3

    # vLLM requirements - vLLM breaks Colab due to reinstalling numpy
    f = requests.get("https://raw.githubusercontent.com/vllm-project/vllm/refs/heads/main/requirements/common.txt").content
    with open("vllm_requirements.txt", "wb") as file:
        file.write(re.sub(rb"(transformers|numpy|xformers)[^\n]{1,}\n", b"", f))
    !pip install -r vllm_requirements.txt

### Unsloth

Load up `Phi-4 14B`, and set parameters

In [5]:
from unsloth import FastLanguageModel, is_bfloat16_supported
import torch
max_seq_length = 512 # Can increase for longer reasoning traces
lora_rank = 16 # Larger rank = smarter, but slower

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/Phi-4",  # Base model
    max_seq_length = max_seq_length,
    load_in_4bit = True,
    fast_inference = True,
    max_lora_rank = lora_rank,
    gpu_memory_utilization = 0.7,
    lora_weights = "JunaidSadiq/Phi4_scholar",
)

model = FastLanguageModel.get_peft_model(
    model,
    r = lora_rank, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["gate_proj", "up_proj", "down_proj",],
    lora_alpha = lora_rank,
    use_gradient_checkpointing = "unsloth", # Enable long context finetuning
    random_state = 3407,
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
INFO 06-08 19:58:40 [importing.py:53] Triton module has been replaced with a placeholder.
INFO 06-08 19:58:40 [__init__.py:239] Automatically detected platform cuda.
Unsloth: Switching from Unsloth dynamic quant to normal quant since
we do not yet support fast inference for unsloth/phi-4-unsloth-bnb-4bit
==((====))==  Unsloth 2025.6.1: Fast Llama patching. Transformers: 4.51.3. vLLM: 0.8.5.post1.
   \\   /|    NVIDIA L4. Num GPUs = 1. Max memory: 22.161 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 8.9. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = TRUE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


config.json:   0%|          | 0.00/1.40k [00:00<?, ?B/s]

Unsloth: vLLM loading unsloth/phi-4-bnb-4bit with actual GPU utilization = 69.29%
Unsloth: Your GPU has CUDA compute capability 8.9 with VRAM = 22.16 GB.
Unsloth: Using conservativeness = 1.0. Chunked prefill tokens = 512. Num Sequences = 192.
Unsloth: vLLM's KV Cache can use up to 5.39 GB. Also swap space = 6 GB.
INFO 06-08 19:59:06 [config.py:717] This model supports multiple tasks: {'classify', 'embed', 'score', 'reward', 'generate'}. Defaulting to 'generate'.
INFO 06-08 19:59:06 [config.py:2003] Chunked prefill is enabled with max_num_batched_tokens=512.
Unsloth: vLLM Bitsandbytes config using kwargs = {'load_in_8bit': False, 'load_in_4bit': True, 'bnb_4bit_compute_dtype': 'bfloat16', 'bnb_4bit_quant_storage': 'uint8', 'bnb_4bit_quant_type': 'nf4', 'bnb_4bit_use_double_quant': True, 'llm_int8_enable_fp32_cpu_offload': False, 'llm_int8_has_fp16_weight': False, 'llm_int8_skip_modules': ['lm_head', 'multi_modal_projector', 'merger', 'modality_projection'], 'llm_int8_threshold': 6.0}


tokenizer_config.json:   0%|          | 0.00/18.0k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.61M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/917k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/7.15M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/170 [00:00<?, ?B/s]

INFO 06-08 19:59:12 [core.py:58] Initializing a V1 LLM engine (v0.8.5.post1) with config: model='unsloth/phi-4-bnb-4bit', speculative_config=None, tokenizer='unsloth/phi-4-bnb-4bit', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.bfloat16, max_seq_len=512, download_dir=None, load_format=LoadFormat.BITSANDBYTES, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=bitsandbytes, enforce_eager=False, kv_cache_dtype=auto,  device_config=cuda:0, decoding_config=DecodingConfig(guided_decoding_backend='auto', reasoning_backend=None), observability_config=ObservabilityConfig(show_hidden_metrics=False, otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=unsloth/phi-4-bnb-4bit, num_scheduler_steps=1, multi_step_stream_outputs=True, enable_prefix_caching=True, chunked_prefill_enab

model-00001-of-00002.safetensors:   0%|          | 0.00/4.98G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/4.11G [00:00<?, ?B/s]

INFO 06-08 19:59:49 [weight_utils.py:281] Time spent downloading weights for unsloth/phi-4-bnb-4bit: 33.785675 seconds


model.safetensors.index.json:   0%|          | 0.00/165k [00:00<?, ?B/s]

Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


INFO 06-08 19:59:55 [punica_selector.py:18] Using PunicaWrapperGPU.
INFO 06-08 19:59:56 [gpu_model_runner.py:1347] Model loading took 8.6253 GiB and 41.613045 seconds
INFO 06-08 20:00:19 [backends.py:420] Using cache directory: /root/.cache/vllm/torch_compile_cache/84c2534a2d/rank_0_0 for vLLM's torch.compile
INFO 06-08 20:00:19 [backends.py:430] Dynamo bytecode transform time: 22.79 s


Inductor Compilation: 100%|██████████| 6/6 [00:00<00:00,  6.43it/s, triton_poi_fused_cat_5]


INFO 06-08 20:00:25 [backends.py:136] Cache the graph of shape None for later use


Inductor Compilation: 100%|██████████| 5/5 [00:00<00:00, 12.78it/s, triton_red_fused__to_copy_add_mean_mul_pow_rsqrt_4]

INFO 06-08 20:01:29 [backends.py:148] Compiling a graph for general shape takes 66.66 s


INFO 06-08 20:03:48 [monitor.py:33] torch.compile takes 89.46 s in total
INFO 06-08 20:03:52 [kv_cache_utils.py:634] GPU KV cache size: 29,920 tokens
INFO 06-08 20:03:52 [kv_cache_utils.py:637] Maximum concurrency for 512 tokens per request: 58.44x
INFO 06-08 20:05:50 [gpu_model_runner.py:1686] Graph capturing finished in 118 secs, took 1.80 GiB
INFO 06-08 20:05:50 [core.py:159] init engine (profile, create kv cache, warmup model) took 354.33 seconds
Unsloth: Just some info: will skip parsing ['pre_feedforward_layernorm', 'post_feedforward_layernorm', 'k_norm', 'q_norm']
Unsloth: Just some info: will skip parsing ['pre_feedforward_layernorm', 'post_feedforward_layernorm', 'k_norm', 'q_norm']


tokenizer_config.json:   0%|          | 0.00/18.0k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.61M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/917k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/7.15M [00:00<?, ?B/s]

Not an error, but Unsloth cannot patch Attention layers with our manual autograd engine since either LoRA adapters
are not enabled or a bias term (like in Qwen) is used.
Not an error, but Unsloth cannot patch O projection layer with our manual autograd engine since either LoRA adapters
are not enabled or a bias term (like in Qwen) is used.
Unsloth 2025.6.1 patched 40 layers with 0 QKV layers, 0 O layers and 40 MLP layers.


### Data Prep
<a name="Data"></a>

We directly leverage [@willccbb](https://gist.github.com/willccbb/4676755236bb08cab5f4e54a0475d6fb) for data prep and all reward functions. You are free to create your own!

In [6]:
import re
from datasets import load_dataset, Dataset

# Load and prep dataset
SYSTEM_PROMPT = """
Respond in the following format:
<reasoning>
...
</reasoning>
<answer>
...
</answer>
"""

XML_COT_FORMAT = """\
<reasoning>
{reasoning}
</reasoning>
<answer>
{answer}
</answer>
"""

def extract_xml_answer(text: str) -> str:
    answer = text.split("<answer>")[-1]
    answer = answer.split("</answer>")[0]
    return answer.strip()

def extract_hash_answer(text: str) -> str | None:
    if "####" not in text:
        return None
    return text.split("####")[1].strip()

# uncomment middle messages for 1-shot prompting
def get_gsm8k_questions(split = "train") -> Dataset:
    data = load_dataset('openai/gsm8k', 'main')[split] # type: ignore
    data = data.map(lambda x: { # type: ignore
        'prompt': [
            {'role': 'system', 'content': SYSTEM_PROMPT},
            {'role': 'user', 'content': x['question']}
        ],
        'answer': extract_hash_answer(x['answer'])
    }) # type: ignore
    return data # type: ignore

dataset = get_gsm8k_questions()

# Reward functions
def correctness_reward_func(prompts, completions, answer, **kwargs) -> list[float]:
    responses = [completion[0]['content'] for completion in completions]
    q = prompts[0][-1]['content']
    extracted_responses = [extract_xml_answer(r) for r in responses]
    print('-'*20, f"Question:\n{q}", f"\nAnswer:\n{answer[0]}", f"\nResponse:\n{responses[0]}", f"\nExtracted:\n{extracted_responses[0]}")
    return [2.0 if r == a else 0.0 for r, a in zip(extracted_responses, answer)]

def int_reward_func(completions, **kwargs) -> list[float]:
    responses = [completion[0]['content'] for completion in completions]
    extracted_responses = [extract_xml_answer(r) for r in responses]
    return [0.5 if r.isdigit() else 0.0 for r in extracted_responses]

def strict_format_reward_func(completions, **kwargs) -> list[float]:
    """Reward function that checks if the completion has a specific format."""
    pattern = r"^<reasoning>\n.*?\n</reasoning>\n<answer>\n.*?\n</answer>\n$"
    responses = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, r) for r in responses]
    return [0.5 if match else 0.0 for match in matches]

def soft_format_reward_func(completions, **kwargs) -> list[float]:
    """Reward function that checks if the completion has a specific format."""
    pattern = r"<reasoning>.*?</reasoning>\s*<answer>.*?</answer>"
    responses = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, r) for r in responses]
    return [0.5 if match else 0.0 for match in matches]

def count_xml(text) -> float:
    count = 0.0
    if text.count("<reasoning>\n") == 1:
        count += 0.125
    if text.count("\n</reasoning>\n") == 1:
        count += 0.125
    if text.count("\n<answer>\n") == 1:
        count += 0.125
        count -= len(text.split("\n</answer>\n")[-1])*0.001
    if text.count("\n</answer>") == 1:
        count += 0.125
        count -= (len(text.split("\n</answer>")[-1]) - 1)*0.001
    return count

def xmlcount_reward_func(completions, **kwargs) -> list[float]:
    contents = [completion[0]["content"] for completion in completions]
    return [count_xml(c) for c in contents]

README.md:   0%|          | 0.00/7.94k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/2.31M [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/419k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/7473 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1319 [00:00<?, ? examples/s]

Map:   0%|          | 0/7473 [00:00<?, ? examples/s]

<a name="Train"></a>
### Train the model

Now set up GRPO Trainer and all configurations!

In [7]:
from trl import GRPOConfig, GRPOTrainer
training_args = GRPOConfig(
    use_vllm = True, # use vLLM for fast inference!
    learning_rate = 5e-6,
    adam_beta1 = 0.9,
    adam_beta2 = 0.99,
    weight_decay = 0.1,
    warmup_ratio = 0.1,
    lr_scheduler_type = "cosine",
    optim = "paged_adamw_8bit",
    logging_steps = 1,
    bf16 = is_bfloat16_supported(),
    fp16 = not is_bfloat16_supported(),
    per_device_train_batch_size = 1,
    gradient_accumulation_steps = 1, # Increase to 4 for smoother training
    num_generations = 6, # Decrease if out of memory
    max_prompt_length = 256,
    max_completion_length = 200,
    # num_train_epochs = 1, # Set to 1 for a full training run
    max_steps = 100,
    save_steps = 250,
    max_grad_norm = 0.1,
    report_to = "wandb", # Can use Weights & Biases
    output_dir = "outputs",
)

Unsloth: We now expect `per_device_train_batch_size` to be a multiple of `num_generations`.
We will change the batch size of 1 to the `num_generations` of 6


And let's run the trainer! If you scroll up, you'll see a table of rewards. The goal is to see the `reward` column increase!

You might have to wait 150 to 200 steps for any action. You'll probably get 0 reward for the first 100 steps. Please be patient!

| Step | Training Loss | reward    | reward_std | completion_length | kl       |
|------|---------------|-----------|------------|-------------------|----------|
| 1    | 0.000000      | 0.125000  | 0.000000   | 200.000000        | 0.000000 |
| 2    | 0.000000      | 0.072375  | 0.248112   | 200.000000        | 0.000000 |
| 3    | 0.000000      | -0.079000 | 0.163776   | 182.500000        | 0.000005 |


In [8]:
trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [
        xmlcount_reward_func,
        soft_format_reward_func,
        strict_format_reward_func,
        int_reward_func,
        correctness_reward_func,
    ],
    args = training_args,
    train_dataset = dataset,
)
trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 7,473 | Num Epochs = 1 | Total steps = 100
O^O/ \_/ \    Batch size per device = 6 | Gradient accumulation steps = 1
\        /    Data Parallel GPUs = 1 | Total batch size (6 x 1 x 1) = 6
 "-____-"     Trainable parameters = 44,236,800/7,888,000,000 (0.56% trained)
wandb: WARNING The `run_name` is currently set to the same value as `TrainingArguments.output_dir`. If this was not intended, please specify a different run name by setting the `TrainingArguments.run_name` parameter.


<IPython.core.display.Javascript object>

wandb: Logging into wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: You can find your API key in your browser here: https://wandb.ai/authorize?ref=models
wandb: Paste an API key from your profile and hit enter:

 ··········


wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: junaidsadiq024 (junaidsadiq024-tampere-university) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


-------------------- Question:
A concert ticket costs $40. Mr. Benson bought 12 tickets and received a 5% discount for every ticket bought that exceeds 10. How much did Mr. Benson pay in all? 
Answer:
476 
Response:
<reasoning>
1. **Determine the number of tickets that receive the discount:**  
   Mr. Benson bought 12 tickets. The discount applies to every ticket exceeding 10 tickets. Therefore, 12 - 10 = 2 tickets receive the 5% discount.

2. **Calculate the cost of non-discounted tickets:**  
   The first 10 tickets are sold at full price.  
   Cost per ticket is $40.  
   Therefore, the cost for 10 tickets = 10 tickets * $40/ticket = $400.

3. **Calculate the discounted price for the additional tickets:**  
   Each of the 2 additional tickets receives a 5% discount.  
   The discount per ticket = 5% of $40 = 0.05 * $40 = $2.  
   Therefore, each discounted ticket costs $40 - $2 = $38.  
   The total cost for the 2 discounted tickets = 2 tickets 
Extracted:
<reasoning>
1. **Determine

Step,Training Loss,reward,reward_std,completion_length,kl,rewards / xmlcount_reward_func,rewards / soft_format_reward_func,rewards / strict_format_reward_func,rewards / int_reward_func,rewards / correctness_reward_func
1,0.000000,0.125000,0.000000,200.000000,0.000000,0.125000,0.000000,0.000000,0.000000,0.000000
2,0.000000,0.104167,0.051031,200.000000,0.000000,0.104167,0.000000,0.000000,0.000000,0.000000
3,0.000000,0.125000,0.000000,200.000000,0.000076,0.125000,0.000000,0.000000,0.000000,0.000000
4,0.000000,0.125000,0.000000,200.000000,0.000104,0.125000,0.000000,0.000000,0.000000,0.000000
5,0.000000,-0.233333,0.096633,196.166672,0.000087,-0.233333,0.000000,0.000000,0.000000,0.000000
6,0.000000,0.125000,0.000000,200.000000,0.000051,0.125000,0.000000,0.000000,0.000000,0.000000
7,0.000000,0.104167,0.051031,200.000000,0.000192,0.104167,0.000000,0.000000,0.000000,0.000000
8,0.000000,0.241000,1.061998,188.166672,0.000199,-0.175667,0.000000,0.000000,0.083333,0.333333
9,0.000000,0.061167,0.223267,200.000000,0.000114,0.061167,0.000000,0.000000,0.000000,0.000000
10,0.000000,0.125000,0.000000,200.000000,0.000088,0.125000,0.000000,0.000000,0.000000,0.000000


-------------------- Question:
Jane is trying to decide whether to buy a house or a trailer. A house costs $480,000 and a trailer costs $120,000. Each loan will be paid in monthly installments over 20 years. How much more is the monthly payment on the house compared to the trailer? 
Answer:
1500 
Response:
<reasoning>
To find the difference in monthly payments between the house and the trailer, we need to calculate the monthly payment for both the house and the trailer using the formula for a fixed-rate mortgage. The formula for the monthly payment \(M\) is:

\[
M = P \frac{r(1+r)^n}{(1+r)^n - 1}
\]

where:
- \(P\) is the loan principal (the amount borrowed).
- \(r\) is the monthly interest rate (annual rate divided by 12).
- \(n\) is the total number of payments (months).

Assuming a fixed interest rate for simplicity (since it's not provided), let's consider a common mortgage rate. We will use 4% annual interest for both loans as an example.

1. **Determine the values for the house:*

TrainOutput(global_step=100, training_loss=1.23237002480181e-05, metrics={'train_runtime': 6012.8903, 'train_samples_per_second': 0.1, 'train_steps_per_second': 0.017, 'total_flos': 0.0, 'train_loss': 1.23237002480181e-05})

<a name="Inference"></a>
### Inference

Now let's try the model we just trained! First, let's first try the model without any GRPO trained:

In [30]:
text = tokenizer.apply_chat_template([
    {"role" : "user", "content" : "Which is bigger? 9.11 or 9.9?"},
], tokenize = False, add_generation_prompt = True)

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 0.8,
    top_p = 0.95,
    max_tokens = 1024,
)
output = model.fast_generate(
    [text],
    sampling_params = sampling_params,
    lora_request = None,
)[0].outputs[0].text

output

Processed prompts:   0%|          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]

' To determine which number is bigger, we compare the numbers digit by digit from left to right.\n\n1. Compare the whole number part:\n   - Both numbers have the whole number part as 9.\n\n2. Compare the tenths place:\n   - 9.11 has a 1 in the tenths place.\n   - 9.9 has a 9 in the tenths place.\n\nSince 1 is less than 9, we can already determine that 9.11 is less than 9.9 without needing to compare further digits.\n\nTherefore, 9.9 is bigger than 9.11.<end_working_out><SOLUTION>9.9</SOLUTION>'

And now with the LoRA we just trained with GRPO - we first save the LoRA first!

Now we load the LoRA and test:

In [31]:
text = tokenizer.apply_chat_template([
    {"role" : "system", "content" : SYSTEM_PROMPT},
    {"role" : "user", "content" : "Which is bigger? 9.11 or 9.9?"},
], tokenize = False, add_generation_prompt = True)

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 0.8,
    top_p = 0.95,
    max_tokens = 1024,
)
output = model.fast_generate(
    text,
    sampling_params = sampling_params,
    lora_request = model.load_lora("grpo_saved_lora"),
)[0].outputs[0].text

output

Processed prompts:   0%|          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]

"Let's compare the numbers 9.11 and 9.9.\n\nFirst, look at the whole number part of each number:\n- The whole number part of 9.11 is 9.\n- The whole number part of 9.9 is also 9.\n\nSince the whole number parts are equal, we need to compare the decimal parts:\n- The decimal part of 9.11 is .11.\n- The decimal part of 9.9 is .9.\n\nTo compare .11 and .9, we can think of them as:\n- .11 is the same as 11/100.\n- .9 is the same as 90/100.\n\nClearly, 90/100 is greater than 11/100. Therefore, .9 is greater than .11.\n\nSo, 9.9 is greater than 9.11.</start_working_out>"

In [32]:
print(output)

Let's compare the numbers 9.11 and 9.9.

First, look at the whole number part of each number:
- The whole number part of 9.11 is 9.
- The whole number part of 9.9 is also 9.

Since the whole number parts are equal, we need to compare the decimal parts:
- The decimal part of 9.11 is .11.
- The decimal part of 9.9 is .9.

To compare .11 and .9, we can think of them as:
- .11 is the same as 11/100.
- .9 is the same as 90/100.

Clearly, 90/100 is greater than 11/100. Therefore, .9 is greater than .11.

So, 9.9 is greater than 9.11.</start_working_out>


Our reasoning model is much better - it's not always correct, since we only trained it for an hour or so - it'll be better if we extend the sequence length and train for longer!

In [33]:
model.save_lora("grpo_saved_lora")

In [35]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if True: model.push_to_hub_merged("JunaidSadiq/phi_reasoning", tokenizer, save_method = "lora", token = "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")

No files have been modified since last commit. Skipping to prevent empty commit.


Found HuggingFace hub cache directory: /root/.cache/huggingface/hub
Checking cache directory for required files...
Cache check failed: model-00001-of-00006.safetensors not found in local cache.
Not all required files found in cache. Will proceed with downloading.


model.safetensors.index.json:   0%|          | 0.00/29.9k [00:00<?, ?B/s]

No files have been modified since last commit. Skipping to prevent empty commit.
Unsloth: Merging weights into 16bit:   0%|          | 0/6 [00:00<?, ?it/s]

model-00001-of-00006.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

  0%|          | 0/1 [00:00<?, ?it/s]

model-00001-of-00006.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

Unsloth: Merging weights into 16bit:  17%|█▋        | 1/6 [01:40<08:22, 100.43s/it]

model-00002-of-00006.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

  0%|          | 0/1 [00:00<?, ?it/s]

model-00002-of-00006.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

Unsloth: Merging weights into 16bit:  33%|███▎      | 2/6 [03:23<06:47, 101.77s/it]

model-00003-of-00006.safetensors:   0%|          | 0.00/4.90G [00:00<?, ?B/s]

  0%|          | 0/1 [00:00<?, ?it/s]

model-00003-of-00006.safetensors:   0%|          | 0.00/4.90G [00:00<?, ?B/s]

Unsloth: Merging weights into 16bit:  50%|█████     | 3/6 [05:01<05:00, 100.23s/it]

model-00004-of-00006.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

  0%|          | 0/1 [00:00<?, ?it/s]

model-00004-of-00006.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

Unsloth: Merging weights into 16bit:  67%|██████▋   | 4/6 [06:36<03:16, 98.16s/it] 

model-00005-of-00006.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

  0%|          | 0/1 [00:00<?, ?it/s]

model-00005-of-00006.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

Unsloth: Merging weights into 16bit:  83%|████████▎ | 5/6 [08:32<01:44, 104.69s/it]

model-00006-of-00006.safetensors:   0%|          | 0.00/4.62G [00:00<?, ?B/s]

  0%|          | 0/1 [00:00<?, ?it/s]

model-00006-of-00006.safetensors:   0%|          | 0.00/4.62G [00:00<?, ?B/s]

Unsloth: Merging weights into 16bit: 100%|██████████| 6/6 [10:02<00:00, 100.43s/it]


Additional Reasoning for Coding with Nvidia/OpenReasoning

In [15]:
reasoning_start = "<start_working_out>" # Acts as <think>
reasoning_end   = "<end_working_out>"   # Acts as </think>
solution_start  = "<SOLUTION>"
solution_end    = "</SOLUTION>"

system_prompt = \
f"""You are given a problem.
Think about the problem and provide your working out.
Place it between {reasoning_start} and {reasoning_end}.
Then, provide your solution between {solution_start}{solution_end}"""
system_prompt

'You are given a problem.\nThink about the problem and provide your working out.\nPlace it between <start_working_out> and <end_working_out>.\nThen, provide your solution between <SOLUTION></SOLUTION>'

In [16]:
chat_template = \
    "{% if messages[0]['role'] == 'system' %}"\
        "{{ messages[0]['content'] + eos_token }}"\
        "{% set loop_messages = messages[1:] %}"\
    "{% else %}"\
        "{{ '{system_prompt}' + eos_token }}"\
        "{% set loop_messages = messages %}"\
    "{% endif %}"\
    "{% for message in loop_messages %}"\
        "{% if message['role'] == 'user' %}"\
            "{{ message['content'] }}"\
        "{% elif message['role'] == 'assistant' %}"\
            "{{ message['content'] + eos_token }}"\
        "{% endif %}"\
    "{% endfor %}"\
    "{% if add_generation_prompt %}{{ '{reasoning_start}' }}"\
    "{% endif %}"

# Replace with out specific template:
chat_template = chat_template\
    .replace("'{system_prompt}'",   f"'{system_prompt}'")\
    .replace("'{reasoning_start}'", f"'{reasoning_start}'")
tokenizer.chat_template = chat_template

In [17]:
tokenizer.apply_chat_template([
    {"role" : "user", "content" : "What is 1+1?"},
    {"role" : "assistant", "content" : f"{reasoning_start}I think it's 2.{reasoning_end}{solution_start}2{solution_end}"},
    {"role" : "user", "content" : "What is 2+2?"},
], tokenize = False, add_generation_prompt = True)

"You are given a problem.\nThink about the problem and provide your working out.\nPlace it between <start_working_out> and <end_working_out>.\nThen, provide your solution between <SOLUTION></SOLUTION><|im_end|>What is 1+1?<start_working_out>I think it's 2.<end_working_out><SOLUTION>2</SOLUTION><|im_end|>What is 2+2?<start_working_out>"

In [20]:
from datasets import load_dataset
import pandas as pd
import numpy as np
import re

print("Loading Nvidia OpenCodeReasoning dataset...")
# Specify the config name 'split_0' or 'split_1'
# Change split='train' to split='split_0' to match the available split
code_dataset = load_dataset("nvidia/OpenCodeReasoning", name="split_0", split="split_0")
print(f"Dataset size: {len(code_dataset)}")

# Take a subset for efficient training (adjust size as needed)
subset_percentage = 0.1  # 10% of the dataset
subset_size = int(len(code_dataset) * subset_percentage)
code_dataset = code_dataset.select(range(subset_size))
print(f"Using subset size: {len(code_dataset)}")

# Inspect dataset structure
print("\nDataset columns:", code_dataset.column_names)
print("\nSample entry:")
print(code_dataset[0])

Loading Nvidia OpenCodeReasoning dataset...


Resolving data files:   0%|          | 0/30 [00:00<?, ?it/s]

Resolving data files:   0%|          | 0/30 [00:00<?, ?it/s]

Loading dataset shards:   0%|          | 0/30 [00:00<?, ?it/s]

Dataset size: 567850
Using subset size: 56785

Dataset columns: ['id', 'input', 'output', 'source', 'license', 'dataset', 'split', 'difficulty', 'solution']

Sample entry:
{'id': 'c0d4209f929db2b5bf3526a47a2520b0', 'input': 'Problem description.\nVipul is a hardworking super-hero who maintains the bracket ratio of all the strings in the world. Recently he indulged himself in saving the string population so much that he lost his ability for checking brackets (luckily, not permanently ).Being his super-hero friend\xa0help him in his time of hardship.\nInput\n\nThe first line of the input contains an integer T denoting the number of test cases. The description of T test cases follows.\nThe first line of each test case contains a single string S denoting the string to be checked.\n\n\nOutput\n\nFor each test case, output a single line printing "YES" or "NO" (without " " and in uppercase only) , denoting if the brackets in the given string is balanced or not .\n\n\nConstraints\n\n1 ≤ T ≤ 10

In [21]:
def extract_code_answer(solution_text):
    """Extract the final answer from code solution texts.

    Looks for patterns like:
    - Explicit "answer is" or "result is" statements
    - Return statements
    - Final value declarations
    - Final line output
    """
    # Clean up solution text
    solution_text = solution_text.strip()

    # Strategy 1: Look for explicit answer declarations
    answer_patterns = [
        r"(?:answer|result|output|solution)(?:\s+is|\s*=|\s*:)\s*[\"']?([\w\d\.\-\+]+)[\"']?",
        r"(?:return|print|output)\s+[\"']?([\w\d\.\-\+]+)[\"']?",
        r"(?:final\s+(?:value|result|answer))\s*[=:]\s*[\"']?([\w\d\.\-\+]+)[\"']?"
    ]

    for pattern in answer_patterns:
        matches = re.findall(pattern, solution_text, re.IGNORECASE)
        if matches:
            return matches[-1]  # Return the last match, usually most relevant

    # Strategy 2: Extract last line with a number if nothing else worked
    lines = solution_text.split('\n')
    for line in reversed(lines):
        line = line.strip()
        if line and re.search(r'\d+', line):
            # Extract just the numeric part if there's text around it
            num_match = re.search(r'[\d\.\-\+]+', line)
            if num_match:
                return num_match.group(0)

    # Strategy 3: Return the last line as fallback
    for line in reversed(lines):
        line = line.strip()
        if line and len(line) < 50:  # Avoid full code blocks
            return line

    # If all else fails, return a placeholder
    return "N/A"

In [24]:
#reward functions
def match_format_exactly(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # Match if format is seen exactly!
        if match_format.search(response) is not None: score += 3.0
        scores.append(score)
    return scores

def match_format_approximately(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # No need to reward <start_working_out> since we always prepend it!
        score += 0.5 if response.count(reasoning_end) == 1 else -1.0
        score += 0.5 if response.count(solution_start) == 1 else -1.0
        score += 0.5 if response.count(solution_end) == 1 else -1.0
        scores.append(score)
    return scores

def code_logic_reward_func(completions, **kwargs):
    """Reward logical reasoning in code problems"""
    responses = [completion[0]["content"] for completion in completions]
    scores = []

    for response in responses:
        score = 0.0
        reasoning_match = re.search(rf"{reasoning_end}(.*?){solution_start}", response, re.DOTALL)

        if reasoning_match:
            reasoning_text = reasoning_match.group(1).strip()

            # Reward algorithmic thinking keywords
            algo_keywords = ['algorithm', 'complexity', 'time', 'space', 'loop', 'iteration',
                           'recursive', 'condition', 'variable', 'function', 'data structure']
            keyword_score = sum(1 for keyword in algo_keywords if keyword in reasoning_text.lower())
            score += min(keyword_score * 0.1, 0.5)  # Max 0.5 from keywords

            # Reward step-by-step breakdown (check for numbered steps)
            if re.search(r'^\s*\d+\.', reasoning_text, re.MULTILINE):
                score += 0.3

            # Reward code analysis patterns
            analysis_patterns = ['analyze', 'examine', 'consider', 'implement', 'approach']
            if any(pattern in reasoning_text.lower() for pattern in analysis_patterns):
                score += 0.2

        scores.append(min(score, 1.0))
    return scores

def code_correctness_reward_func(prompts, completions, answer, **kwargs):
    """Enhanced correctness for code problems"""
    responses = [completion[0]['content'] for completion in completions]
    extracted_responses = [
        guess.group(1) if (guess := match_format.search(r)) is not None else None
        for r in responses
    ]

    scores = []
    for guess, true_answer in zip(extracted_responses, answer):
        if guess is None:
            scores.append(-1.0)
            continue

        # Exact match gets highest reward
        if guess.strip() == true_answer.strip():
            score = 4.0
        # Match after cleaning whitespace/punctuation
        elif re.sub(r'[\s\.,;\'"]', '', guess).lower() == re.sub(r'[\s\.,;\'"]', '', true_answer).lower():
            score = 3.0
        # Numbers match when extracted
        elif (num_guess := re.search(r'[-+]?\d*\.?\d+', guess)) and \
             (num_answer := re.search(r'[-+]?\d*\.?\d+', true_answer)) and \
             num_guess.group(0) == num_answer.group(0):
            score = 2.5
        # At least some overlap in terms
        else:
            common_terms = set(re.findall(r'\w+', guess.lower())) & set(re.findall(r'\w+', true_answer.lower()))
            if len(common_terms) > 0:
                score = min(1.0, len(common_terms) * 0.2)
            else:
                score = -0.5

        scores.append(score)
    return scores

In [36]:
#
from datasets import load_dataset
import pandas as pd
import numpy as np
import re

print("Loading Nvidia OpenCodeReasoning dataset...")
# Specify the config name 'split_0' or 'split_1'
code_dataset = load_dataset("nvidia/OpenCodeReasoning", name="split_0", split="split_0")
print(f"Dataset size: {len(code_dataset)}")

# Take a subset for efficient training (adjust size as needed)
subset_percentage = 0.1  # 10% of the dataset
subset_size = int(len(code_dataset) * subset_percentage)
code_dataset = code_dataset.select(range(subset_size))
print(f"Using subset size: {len(code_dataset)}")

# Inspect dataset structure
print("\nDataset columns:", code_dataset.column_names)
print("\nSample entry:")
print(code_dataset[0])

# Transform the OpenCodeReasoning dataset to match expected format
def transform_code_dataset(example):
    """Transform OpenCodeReasoning dataset to GRPO format"""
    # Extract problem and solution from the dataset
    problem = example.get('problem', '') or example.get('question', '') or example.get('input', '')
    solution = example.get('solution', '') or example.get('answer', '') or example.get('output', '')

    return {
        "prompt": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": problem},
        ],
        "answer": solution
    }

# Apply transformation
code_dataset = code_dataset.map(transform_code_dataset)

# Define regex pattern for format matching
match_format = re.compile(
    rf"{reasoning_end}.*?{solution_start}(.+?){solution_end}",
    flags = re.MULTILINE | re.DOTALL
)

def match_format_exactly(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # Match if format is seen exactly!
        if match_format.search(response) is not None:
            score += 3.0
        scores.append(score)
    return scores

def match_format_approximately(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # No need to reward <start_working_out> since we always prepend it!
        score += 0.5 if response.count(reasoning_end) == 1 else -1.0
        score += 0.5 if response.count(solution_start) == 1 else -1.0
        score += 0.5 if response.count(solution_end) == 1 else -1.0
        scores.append(score)
    return scores

def code_logic_reward_func(completions, **kwargs):
    """Reward logical reasoning in code problems"""
    responses = [completion[0]["content"] for completion in completions]
    scores = []

    for response in responses:
        score = 0.0
        reasoning_match = re.search(rf"{reasoning_end}(.*?){solution_start}", response, re.DOTALL)

        if reasoning_match:
            reasoning_text = reasoning_match.group(1).strip()

            # Reward algorithmic thinking keywords
            algo_keywords = ['algorithm', 'complexity', 'time', 'space', 'loop', 'iteration',
                           'recursive', 'condition', 'variable', 'function', 'data structure']
            keyword_score = sum(1 for keyword in algo_keywords if keyword in reasoning_text.lower())
            score += min(keyword_score * 0.1, 0.5)  # Max 0.5 from keywords

            # Reward step-by-step breakdown (check for numbered steps)
            if re.search(r'^\s*\d+\.', reasoning_text, re.MULTILINE):
                score += 0.3

            # Reward code analysis patterns
            analysis_patterns = ['analyze', 'examine', 'consider', 'implement', 'approach']
            if any(pattern in reasoning_text.lower() for pattern in analysis_patterns):
                score += 0.2

        scores.append(min(score, 1.0))
    return scores

def code_correctness_reward_func(prompts, completions, answer, **kwargs):
    """Enhanced correctness for code problems"""
    responses = [completion[0]['content'] for completion in completions]
    extracted_responses = [
        guess.group(1) if (guess := match_format.search(r)) is not None else None
        for r in responses
    ]

    scores = []
    for guess, true_answer in zip(extracted_responses, answer):
        if guess is None:
            scores.append(-1.0)
            continue

        # Exact match gets highest reward
        if guess.strip() == true_answer.strip():
            score = 4.0
        # Match after cleaning whitespace/punctuation
        elif re.sub(r'[\s\.,;\'"]', '', guess).lower() == re.sub(r'[\s\.,;\'"]', '', true_answer).lower():
            score = 3.0
        # Numbers match when extracted
        elif (num_guess := re.search(r'[-+]?\d*\.?\d+', guess)) and \
             (num_answer := re.search(r'[-+]?\d*\.?\d+', true_answer)) and \
             num_guess.group(0) == num_answer.group(0):
            score = 2.5
        # At least some overlap in terms
        else:
            common_terms = set(re.findall(r'\w+', guess.lower())) & set(re.findall(r'\w+', true_answer.lower()))
            if len(common_terms) > 0:
                score = min(1.0, len(common_terms) * 0.2)
            else:
                score = -0.5

        scores.append(score)
    return scores


Loading Nvidia OpenCodeReasoning dataset...


Resolving data files:   0%|          | 0/30 [00:00<?, ?it/s]

Resolving data files:   0%|          | 0/30 [00:00<?, ?it/s]

Loading dataset shards:   0%|          | 0/30 [00:00<?, ?it/s]

Dataset size: 567850
Using subset size: 56785

Dataset columns: ['id', 'input', 'output', 'source', 'license', 'dataset', 'split', 'difficulty', 'solution']

Sample entry:
{'id': 'c0d4209f929db2b5bf3526a47a2520b0', 'input': 'Problem description.\nVipul is a hardworking super-hero who maintains the bracket ratio of all the strings in the world. Recently he indulged himself in saving the string population so much that he lost his ability for checking brackets (luckily, not permanently ).Being his super-hero friend\xa0help him in his time of hardship.\nInput\n\nThe first line of the input contains an integer T denoting the number of test cases. The description of T test cases follows.\nThe first line of each test case contains a single string S denoting the string to be checked.\n\n\nOutput\n\nFor each test case, output a single line printing "YES" or "NO" (without " " and in uppercase only) , denoting if the brackets in the given string is balanced or not .\n\n\nConstraints\n\n1 ≤ T ≤ 10

In [ ]:
# Clear memory and restart fresh
import torch
import gc
torch.cuda.empty_cache()
gc.collect()

from datasets import load_dataset, Dataset
import pandas as pd
import numpy as np
import re
from trl import GRPOConfig, GRPOTrainer
from vllm import SamplingParams

print("Loading Nvidia OpenCodeReasoning dataset...")
code_dataset = load_dataset("nvidia/OpenCodeReasoning", name="split_0", split="split_0")
print(f"Dataset size: {len(code_dataset)}")

# Use 1% of the dataset for meaningful training
subset_percentage = 0.01  # 1% of the dataset
subset_size = int(len(code_dataset) * subset_percentage)
code_dataset = code_dataset.select(range(subset_size))
print(f"Using 1% subset size: {len(code_dataset)}")

# Inspect dataset structure
print("\nDataset columns:", code_dataset.column_names)
print("\nSample entry:")
print(code_dataset[0])

# Transform with very aggressive text truncation to fit within token limits
def transform_code_dataset(example):
    """Transform OpenCodeReasoning dataset to GRPO format with aggressive truncation"""
    problem = example.get('problem', '') or example.get('question', '') or example.get('input', '')
    solution = example.get('solution', '') or example.get('answer', '') or example.get('output', '')

    # Very aggressive truncation - approximately 4 chars per token
    # Target: ~200 tokens for problem, ~50 tokens for solution
    if len(problem) > 800:  # ~200 tokens max
        # Try to cut at a sentence boundary
        truncated = problem[:800]
        last_period = truncated.rfind('.')
        last_newline = truncated.rfind('\n')
        cut_point = max(last_period, last_newline)
        if cut_point > 400:  # If we found a good cut point
            problem = problem[:cut_point + 1]
        else:
            problem = problem[:800] + "..."

    if len(solution) > 200:  # ~50 tokens max
        # Try to preserve the final answer
        truncated = solution[:200]
        # Look for common ending patterns
        last_period = truncated.rfind('.')
        if last_period > 100:
            solution = solution[:last_period + 1]
        else:
            solution = solution[:200] + "..."

    return {
        "prompt": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": problem},
        ],
        "answer": solution
    }

# Apply transformation
code_dataset = code_dataset.map(transform_code_dataset)

# Define regex pattern for format matching
match_format = re.compile(
    rf"{reasoning_end}.*?{solution_start}(.+?){solution_end}",
    flags = re.MULTILINE | re.DOTALL
)

def match_format_exactly(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # Match if format is seen exactly!
        if match_format.search(response) is not None:
            score += 3.0
        scores.append(score)
    return scores

def match_format_approximately(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # No need to reward <start_working_out> since we always prepend it!
        score += 0.5 if response.count(reasoning_end) == 1 else -1.0
        score += 0.5 if response.count(solution_start) == 1 else -1.0
        score += 0.5 if response.count(solution_end) == 1 else -1.0
        scores.append(score)
    return scores

def code_logic_reward_func(completions, **kwargs):
    """Reward logical reasoning in code problems"""
    responses = [completion[0]["content"] for completion in completions]
    scores = []

    for response in responses:
        score = 0.0
        reasoning_match = re.search(rf"{reasoning_end}(.*?){solution_start}", response, re.DOTALL)

        if reasoning_match:
            reasoning_text = reasoning_match.group(1).strip()

            # Reward algorithmic thinking keywords
            algo_keywords = ['algorithm', 'complexity', 'time', 'space', 'loop', 'iteration',
                           'recursive', 'condition', 'variable', 'function', 'data structure']
            keyword_score = sum(1 for keyword in algo_keywords if keyword in reasoning_text.lower())
            score += min(keyword_score * 0.1, 0.5)  # Max 0.5 from keywords

            # Reward step-by-step breakdown (check for numbered steps)
            if re.search(r'^\s*\d+\.', reasoning_text, re.MULTILINE):
                score += 0.3

            # Reward code analysis patterns
            analysis_patterns = ['analyze', 'examine', 'consider', 'implement', 'approach']
            if any(pattern in reasoning_text.lower() for pattern in analysis_patterns):
                score += 0.2

        scores.append(min(score, 1.0))
    return scores

def code_correctness_reward_func(prompts, completions, answer, **kwargs):
    """Enhanced correctness for code problems"""
    responses = [completion[0]['content'] for completion in completions]
    extracted_responses = [
        guess.group(1) if (guess := match_format.search(r)) is not None else None
        for r in responses
    ]

    scores = []
    for guess, true_answer in zip(extracted_responses, answer):
        if guess is None:
            scores.append(-1.0)
            continue

        # Exact match gets highest reward
        if guess.strip() == true_answer.strip():
            score = 4.0
        # Match after cleaning whitespace/punctuation
        elif re.sub(r'[\s\.,;\'"]', '', guess).lower() == re.sub(r'[\s\.,;\'"]', '', true_answer).lower():
            score = 3.0
        # Numbers match when extracted
        elif (num_guess := re.search(r'[-+]?\d*\.?\d+', guess)) and \
             (num_answer := re.search(r'[-+]?\d*\.?\d+', true_answer)) and \
             num_guess.group(0) == num_answer.group(0):
            score = 2.5
        # At least some overlap in terms
        else:
            common_terms = set(re.findall(r'\w+', guess.lower())) & set(re.findall(r'\w+', true_answer.lower()))
            if len(common_terms) > 0:
                score = min(1.0, len(common_terms) * 0.2)
            else:
                score = -0.5

        scores.append(score)
    return scores

# Calculate token lengths and filter aggressively
tokenized = code_dataset.map(
    lambda x: {"tokens": tokenizer.apply_chat_template(x["prompt"], add_generation_prompt=True, tokenize=True)},
    batched=True,
)
tokenized = tokenized.map(lambda x: {"L": len(x["tokens"])})

print(f"Token length statistics:")
print(f"Min: {min(tokenized['L'])}")
print(f"Max: {max(tokenized['L'])}")
print(f"Mean: {np.mean(tokenized['L']):.2f}")
print(f"75th percentile: {np.quantile(tokenized['L'], 0.75):.0f}")
print(f"90th percentile: {np.quantile(tokenized['L'], 0.9):.0f}")

# Use 60th percentile to keep more samples but still be safe
maximum_length = int(np.quantile(tokenized["L"], 0.6))
print("Max Length (60th percentile) = ", maximum_length)

# Filter to keep prompts under a strict limit
max_allowed_prompt_length = min(maximum_length, 280)  # Strict limit
filtered_indices = np.where(np.array(tokenized["L"]) <= max_allowed_prompt_length)[0]
code_dataset = code_dataset.select(filtered_indices)
print(f"Filtered dataset size: {len(code_dataset)} samples")

if len(code_dataset) < 50:
    print(f"Warning: Only {len(code_dataset)} samples after filtering. This might be too few for training.")
    # Relax the filter slightly if we have too few samples
    max_allowed_prompt_length = min(320, max_seq_length - 180)
    filtered_indices = np.where(np.array(tokenized["L"]) <= max_allowed_prompt_length)[0]
    code_dataset = code_dataset.select(filtered_indices)
    print(f"Relaxed filtering gives us {len(code_dataset)} samples")

# Recalculate after filtering
tokenized_filtered = code_dataset.map(
    lambda x: {"tokens": tokenizer.apply_chat_template(x["prompt"], add_generation_prompt=True, tokenize=True)},
    batched=True,
)
tokenized_filtered = tokenized_filtered.map(lambda x: {"L": len(x["tokens"])})
final_max_length = max(tokenized_filtered["L"])
print("Final max length after filtering = ", final_max_length)

del tokenized, tokenized_filtered

# Set conservative but reasonable prompt and completion lengths
max_prompt_length = min(final_max_length + 10, 300)  # Small buffer
max_completion_length = max_seq_length - max_prompt_length - 20  # Extra safety margin

# Ensure we have reasonable values
print(f"max_seq_length: {max_seq_length}")
print(f"max_prompt_length: {max_prompt_length}")
print(f"max_completion_length: {max_completion_length}")

# Define vllm_sampling_params
vllm_sampling_params = SamplingParams(
    min_p = 0.1,
    top_p = 1.0,
    top_k = -1,
    seed = 3407,
    stop = [tokenizer.eos_token],
    include_stop_str_in_output = True,
    max_tokens = max_completion_length,
)

# GRPO Configuration with reasonable settings for 1% dataset
training_args_code = GRPOConfig(
    vllm_sampling_params = vllm_sampling_params,
    temperature = 0.8,
    learning_rate = 3e-6,
    weight_decay = 0.05,
    warmup_ratio = 0.1,
    lr_scheduler_type = "cosine",
    optim = "adamw_8bit",
    logging_steps = 1,
    per_device_train_batch_size = 1,
    gradient_accumulation_steps = 2,  # Reasonable accumulation
    num_generations = 2,  # Keep memory usage reasonable
    max_prompt_length = max_prompt_length,
    max_completion_length = max_completion_length,
    max_steps = 200,  # More steps for 1% dataset
    save_steps = 50,
    report_to = "wandb",
    output_dir = "outputs_code_reasoning_1pct",
    run_name = "phi4-CodeReasoning-1percent"
)

# Initialize and train the GRPO trainer for code reasoning
trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [
        match_format_exactly,
        match_format_approximately,
        code_logic_reward_func,
        code_correctness_reward_func,
    ],
    args = training_args_code,
    train_dataset = code_dataset,
)

print(f"Starting GRPO training with {len(code_dataset)} samples (1% of dataset)...")
print(f"Training for {training_args_code.max_steps} steps")
trainer.train()

Loading Nvidia OpenCodeReasoning dataset...


Resolving data files:   0%|          | 0/30 [00:00<?, ?it/s]

Resolving data files:   0%|          | 0/30 [00:00<?, ?it/s]

Loading dataset shards:   0%|          | 0/30 [00:00<?, ?it/s]

Dataset size: 567850
Using 1% subset size: 5678

Dataset columns: ['id', 'input', 'output', 'source', 'license', 'dataset', 'split', 'difficulty', 'solution']

Sample entry:
{'id': 'c0d4209f929db2b5bf3526a47a2520b0', 'input': 'Problem description.\nVipul is a hardworking super-hero who maintains the bracket ratio of all the strings in the world. Recently he indulged himself in saving the string population so much that he lost his ability for checking brackets (luckily, not permanently ).Being his super-hero friend\xa0help him in his time of hardship.\nInput\n\nThe first line of the input contains an integer T denoting the number of test cases. The description of T test cases follows.\nThe first line of each test case contains a single string S denoting the string to be checked.\n\n\nOutput\n\nFor each test case, output a single line printing "YES" or "NO" (without " " and in uppercase only) , denoting if the brackets in the given string is balanced or not .\n\n\nConstraints\n\n1 ≤ T ≤ 

Map:   0%|          | 0/5678 [00:00<?, ? examples/s]

Map:   0%|          | 0/5678 [00:00<?, ? examples/s]

Map:   0%|          | 0/5678 [00:00<?, ? examples/s]

Token length statistics:
Min: 58
Max: 461
Mean: 230.51
75th percentile: 248
90th percentile: 274
Max Length (60th percentile) =  235
Filtered dataset size: 3472 samples


Map:   0%|          | 0/3472 [00:00<?, ? examples/s]

Map:   0%|          | 0/3472 [00:00<?, ? examples/s]

Final max length after filtering =  235
max_seq_length: 1024
max_prompt_length: 245
max_completion_length: 759
Unsloth: We now expect `per_device_train_batch_size` to be a multiple of `num_generations`.
We will change the batch size of 1 to the `num_generations` of 2
Starting GRPO training with 3472 samples (1% of dataset)...
Training for 200 steps


==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 3,472 | Num Epochs = 1 | Total steps = 200
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 2
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 2 x 1) = 4
 "-____-"     Trainable parameters = 44,236,800/7,888,000,000 (0.56% trained)


Step,Training Loss,reward,reward_std,completion_length,kl,rewards / match_format_exactly,rewards / match_format_approximately,rewards / code_logic_reward_func,rewards / code_correctness_reward_func
1,0.000000,-2.500000,2.121320,296.000000,0.000365,0.000000,-1.500000,0.000000,-1.000000
2,0.000000,-2.125000,0.530330,277.000000,0.000268,0.000000,-1.125000,0.000000,-1.000000
3,0.000000,-3.250000,0.000000,265.500000,0.000382,0.000000,-2.250000,0.000000,-1.000000
4,0.000000,-2.875000,1.590990,284.000000,0.000206,0.000000,-1.875000,0.000000,-1.000000
5,0.000000,-2.875000,0.530330,304.000000,0.000538,0.000000,-1.875000,0.000000,-1.000000
6,0.000000,-2.500000,1.060660,227.000000,0.000340,0.000000,-1.500000,0.000000,-1.000000
7,-0.000000,-2.500000,1.060660,269.000000,0.000326,0.000000,-1.500000,0.000000,-1.000000
8,0.000000,-4.000000,0.000000,303.000000,0.000332,0.000000,-3.000000,0.000000,-1.000000
9,-0.000000,-3.250000,1.060660,306.000000,0.000274,0.000000,-2.250000,0.000000,-1.000000
10,-0.000000,-0.125000,3.358757,310.000000,0.000577,0.750000,-0.375000,0.000000,-0.500000


Step,Training Loss,reward,reward_std,completion_length,kl,rewards / match_format_exactly,rewards / match_format_approximately,rewards / code_logic_reward_func,rewards / code_correctness_reward_func
1,0.000000,-2.500000,2.121320,296.000000,0.000365,0.000000,-1.500000,0.000000,-1.000000
2,0.000000,-2.125000,0.530330,277.000000,0.000268,0.000000,-1.125000,0.000000,-1.000000
3,0.000000,-3.250000,0.000000,265.500000,0.000382,0.000000,-2.250000,0.000000,-1.000000
4,0.000000,-2.875000,1.590990,284.000000,0.000206,0.000000,-1.875000,0.000000,-1.000000
5,0.000000,-2.875000,0.530330,304.000000,0.000538,0.000000,-1.875000,0.000000,-1.000000
6,0.000000,-2.500000,1.060660,227.000000,0.000340,0.000000,-1.500000,0.000000,-1.000000
7,-0.000000,-2.500000,1.060660,269.000000,0.000326,0.000000,-1.500000,0.000000,-1.000000
8,0.000000,-4.000000,0.000000,303.000000,0.000332,0.000000,-3.000000,0.000000,-1.000000
9,-0.000000,-3.250000,1.060660,306.000000,0.000274,0.000000,-2.250000,0.000000,-1.000000
10,-0.000000,-0.125000,3.358757,310.000000,0.000577,0.750000,-0.375000,0.000000,-0.500000


In [ ]:
from datasets import load_dataset
import pandas as pd
import numpy as np

dataset = load_dataset("unsloth/OpenMathReasoning-mini", split = "cot")
dataset = dataset.to_pandas()[
    ["expected_answer", "problem", "generated_solution"]
]

# Try converting to number - if not, replace with NaN
is_number = pd.to_numeric(pd.Series(dataset["expected_answer"]), errors = "coerce").notnull()
# Select only numbers
dataset = dataset.iloc[np.where(is_number)[0]]

dataset

In [ ]:
def format_dataset(x):
    expected_answer = x["expected_answer"]
    problem = x["problem"]

    # Remove generated <think> and </think>
    thoughts = x["generated_solution"]
    thoughts = thoughts.replace("<think>", "").replace("</think>", "")

    # Strip newlines on left and right
    thoughts = thoughts.strip()
    # Add our custom formatting
    final_prompt = \
        reasoning_start + thoughts + reasoning_end + \
        solution_start + expected_answer + solution_end
    return [
        {"role" : "system",    "content" : system_prompt},
        {"role" : "user",      "content" : problem},
        {"role" : "assistant", "content" : final_prompt},
    ]

dataset["Messages"] = dataset.apply(format_dataset, axis = 1)

In [ ]:
tokenizer.apply_chat_template(dataset["Messages"][0], tokenize = False)

In [ ]:
dataset["N"] = dataset["Messages"].apply(lambda x: len(tokenizer.apply_chat_template(x)))

dataset = dataset.loc[dataset["N"] <= max_seq_length/2].copy()
dataset.shape

In [ ]:
from datasets import Dataset

dataset["text"] = tokenizer.apply_chat_template(dataset["Messages"].values.tolist(), tokenize = False)
dataset = Dataset.from_pandas(dataset)
dataset

In [ ]:
from trl import SFTTrainer, SFTConfig
trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    args = SFTConfig(
        dataset_text_field = "text",
        per_device_train_batch_size = 1,
        gradient_accumulation_steps = 1, # Use GA to mimic batch size!
        warmup_steps = 5,
        num_train_epochs = 2, # Set this for 1 full training run.
        learning_rate = 2e-4, # Reduce to 2e-5 for long training runs
        logging_steps = 5,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        report_to = "wandb", # Use this for WandB etc
    ),
)

In [ ]:
trainer.train()

In [ ]:
text = tokenizer.apply_chat_template(
    dataset[0]["Messages"][:2],
    tokenize = False,
    add_generation_prompt = True, # Must add for generation
)

from transformers import TextStreamer
_ = model.generate(
    **tokenizer(text, return_tensors = "pt").to("cuda"),
    temperature = 0,
    max_new_tokens = 1024,
    streamer = TextStreamer(tokenizer, skip_prompt = False),
)

In [ ]:
del dataset
torch.cuda.empty_cache()
import gc
gc.collect()

In [ ]:
model.save_lora("grpo_saved_lora")

In [ ]:
from datasets import load_dataset
dataset = load_dataset("open-r1/DAPO-Math-17k-Processed", "en", split = "train")
dataset

In [ ]:
dataset[0]["prompt"]
dataset[0]["solution"]

In [ ]:
dataset = dataset.map(lambda x: {
    "prompt" : [
        {"role": "system", "content": system_prompt},
        {"role": "user",   "content": x["prompt"]},
    ],
    "answer": extract_hash_answer(x["solution"]),
})
dataset[0]

In [ ]:
import re

# Add optional EOS token matching
solution_end_regex = r"</SOLUTION>[\s]{0,}" + \
    "(?:" + re.escape(tokenizer.eos_token) + ")?"

match_format = re.compile(
    rf"{reasoning_end}.*?"\
    rf"{solution_start}(.+?){solution_end_regex}"\
    rf"[\s]{{0,}}$",
    flags = re.MULTILINE | re.DOTALL
)
match_format

In [ ]:
match_format.findall(
    "Let me think!<end_working_out>"\
    f"<SOLUTION>\n2\n</SOLUTION>",
)

In [ ]:
match_format.findall(
    "<start_working_out>Let me think!<end_working_out>"\
    f"<SOLUTION>  2  </SOLUTION>\n\n",
)

In [ ]:
def match_format_exactly(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # Match if format is seen exactly!
        if match_format.search(response) is not None: score += 3.0
        scores.append(score)
    return scores

In [ ]:
def match_format_approximately(completions, **kwargs):
    scores = []
    for completion in completions:
        score = 0
        response = completion[0]["content"]
        # Count how many keywords are seen - we penalize if too many!
        # If we see 1, then plus some points!

        # No need to reward <start_working_out> since we always prepend it!
        # score += 0.5 if response.count(reasoning_start) == 1 else -1.0
        score += 0.5 if response.count(reasoning_end)   == 1 else -1.0
        score += 0.5 if response.count(solution_start)  == 1 else -1.0
        score += 0.5 if response.count(solution_end)    == 1 else -1.0
        scores.append(score)
    return scores

In [ ]:
def check_answer(prompts, completions, answer, **kwargs):
    question = prompts[0][-1]["content"]
    responses = [completion[0]["content"] for completion in completions]

    extracted_responses = [
        guess.group(1)
        if (guess := match_format.search(r)) is not None else None \
        for r in responses
    ]

    scores = []
    for guess, true_answer in zip(extracted_responses, answer):
        score = 0
        if guess is None:
            scores.append(-2.0)
            continue
        # Correct answer gets 5 points!
        if guess == true_answer:
            score += 5.0
        # Match if spaces are seen, but less reward
        elif guess.strip() == true_answer.strip():
            score += 3.5
        else:
            # We also reward it if the answer is close via ratios!
            # Ie if the answer is within some range, reward it!
            try:
                ratio = float(guess) / float(true_answer)
                if   ratio >= 0.9 and ratio <= 1.1: score += 2.0
                elif ratio >= 0.8 and ratio <= 1.2: score += 1.5
                else: score -= 2.5 # Penalize wrong answers
            except:
                score -= 4.5 # Penalize
        scores.append(score)
    return scores

In [ ]:
match_numbers = re.compile(
    solution_start + r".*?[\s]{0,}([-]?[\d\.\,]{1,})",
    flags = re.MULTILINE | re.DOTALL
)
print(match_numbers.findall("<SOLUTION>  0.34  </SOLUTION>"))
print(match_numbers.findall("<SOLUTION>  123,456  </SOLUTION>"))
print(match_numbers.findall("<SOLUTION>  -0.234  </SOLUTION>"))
print(match_numbers.findall("<SOLUTION>17</SOLUTION>"))

In [ ]:
global PRINTED_TIMES
PRINTED_TIMES = 0
global PRINT_EVERY_STEPS
PRINT_EVERY_STEPS = 5

def check_numbers(prompts, completions, answer, **kwargs):
    question = prompts[0][-1]["content"]
    responses = [completion[0]["content"] for completion in completions]

    extracted_responses = [
        guess.group(1)
        if (guess := match_numbers.search(r)) is not None else None \
        for r in responses
    ]

    scores = []
    # Print only every few steps
    global PRINTED_TIMES
    global PRINT_EVERY_STEPS
    if PRINTED_TIMES % PRINT_EVERY_STEPS == 0:
        print(
            '*'*20 + f"Question:\n{question}", f"\nAnswer:\n{answer[0]}", f"\nResponse:\n{responses[0]}", f"\nExtracted:\n{extracted_responses[0]}"
        )
    PRINTED_TIMES += 1

    for guess, true_answer in zip(extracted_responses, answer):
        if guess is None:
            scores.append(-2.5)
            continue
        # Convert to numbers
        try:
            true_answer = float(true_answer.strip())
            # Remove commas like in 123,456
            guess       = float(guess.strip().replace(",", ""))
            scores.append(3.5 if guess == true_answer else -1.5)
        except:
            scores.append(0)
            continue
    return scores

In [ ]:
tokenized = dataset.map(
    lambda x: {"tokens" : tokenizer.apply_chat_template(x["prompt"], add_generation_prompt = True, tokenize = True)},
    batched = True,
)
print(tokenizer.decode(tokenized[0]["tokens"]))
tokenized = tokenized.map(lambda x: {"L" : len(x["tokens"])})

import numpy as np
maximum_length = int(np.quantile(tokenized["L"], 0.9))
print("Max Length = ", maximum_length)

# Filter only samples smaller than 90% max length
dataset = dataset.select(np.where(np.array(tokenized["L"]) <= maximum_length)[0])
del tokenized

In [ ]:
max_prompt_length = maximum_length + 1 # + 1 just in case!
max_completion_length = max_seq_length - max_prompt_length

from vllm import SamplingParams
vllm_sampling_params = SamplingParams(
    min_p = 0.1,
    top_p = 1.0,
    top_k = -1,
    seed = 3407,
    stop = [tokenizer.eos_token],
    include_stop_str_in_output = True,
)

from trl import GRPOConfig, GRPOTrainer
training_args = GRPOConfig(
    vllm_sampling_params = vllm_sampling_params,
    temperature = 1.0,
    learning_rate = 5e-6,
    weight_decay = 0.01,
    warmup_ratio = 0.1,
    lr_scheduler_type = "linear",
    optim = "adamw_8bit",
    logging_steps = 1,
    per_device_train_batch_size = 1,
    gradient_accumulation_steps = 1, # Increase to 4 for smoother training
    num_generations = 4, # Decrease if out of memory
    max_prompt_length = max_prompt_length,
    max_completion_length = max_completion_length,
    # num_train_epochs = 1, # Set to 1 for a full training run
    max_steps = 100,
    save_steps = 100,
    report_to = "wandb", # Can use Weights & Biases
    output_dir = "outputs",

    # For optional training + evaluation
    # fp16_full_eval = True,
    # per_device_eval_batch_size = 4,
    # eval_accumulation_steps = 1,
    # eval_strategy = "steps",
    # eval_steps = 1,
)

In [ ]:
# For optional training + evaluation
# new_dataset = dataset.train_test_split(test_size = 0.01)

trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [
        match_format_exactly,
        match_format_approximately,
        check_answer,
        check_numbers,
    ],
    args = training_args,
    train_dataset = dataset,

    # For optional training + evaluation
    # train_dataset = new_dataset["train"],
    # eval_dataset = new_dataset["test"],
)
trainer.train()

In [ ]:
text = "What is the sqrt of 101?"

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 1.0,
    top_k = 50,
    max_tokens = 1024,
)
output = model.fast_generate(
    [text],
    sampling_params = sampling_params,
    lora_request = None,
)[0].outputs[0].text

output

In [ ]:
from safetensors import safe_open

tensors = {}
with safe_open("grpo_saved_lora/adapter_model.safetensors", framework = "pt") as f:
    # Verify both A and B are non zero
    for key in f.keys():
        tensor = f.get_tensor(key)
        n_zeros = (tensor == 0).sum() / tensor.numel()
        assert(n_zeros.item() != tensor.numel())

In [ ]:
messages = [
    {"role": "system", "content": system_prompt},
    {"role": "user",   "content": "What is the sqrt of 101?"},
]

text = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True, # Must add for generation
    tokenize = False,
)
from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 1.0,
    top_k = 50,
    max_tokens = 2048,
)
output = model.fast_generate(
    text,
    sampling_params = sampling_params,
    lora_request = model.load_lora("grpo_saved_lora"),
)[0].outputs[0].text

output

<a name="Save"></a>
### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
